cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/FlowletRuntimeService.java

```diff
@@ -105,48 +109,66 @@ void resume() {
   }
 
   private void initFlowlet() throws InterruptedException {
+    LOG.info("Initializing flowlet: " + flowletContext);
+    TxRunnable runnable = new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        ClassLoader classLoader = setContextCombinedClassLoader();
+        try {
+          flowlet.initialize(flowletContext);
+        } finally {
+          ClassLoaders.setContextClassLoader(classLoader);
+        }
+      }
+    };
     try {
-      dataFabricFacade.createTransactionExecutor().execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() throws Exception {
-          LOG.info("Initializing flowlet: " + flowletContext);
-          ClassLoader classLoader = setContextCombinedClassLoader();
-          try {
-            flowlet.initialize(flowletContext);
-          } finally {
-            ClassLoaders.setContextClassLoader(classLoader);
-          }
-          LOG.info("Flowlet initialized: " + flowletContext);
+      if (TransactionControl.IMPLICIT == Transactions.getTransactionControl(
+        TransactionControl.IMPLICIT, Flowlet.class, flowlet, "initialize", FlowletContext.class)) {
+        try {
+          flowletContext.execute(runnable);
+        } catch (TransactionFailureException e) {
+          throw e.getCause() == null ? e : e.getCause();
         }
-      });
-    } catch (TransactionFailureException e) {
-      Throwable cause = e.getCause() == null ? e : e.getCause();
+      } else {
+        runnable.run(flowletContext);
+      }
+      LOG.info("Flowlet initialized: " + flowletContext);
+    } catch (Throwable cause) {
       LOG.error("Flowlet throws exception during flowlet initialize: " + flowletContext, cause);
```

Comment:
That was actually not a part of this change but existing code. But I can change it, of course.